common/nexus/payload_serializer.go

```diff
@@ -0,0 +1,185 @@
+// The MIT License
+//
+// Copyright (c) 2023 Temporal Technologies Inc.  All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package nexus
+
+import (
+	"errors"
+	"fmt"
+	"maps"
+	"mime"
+
+	"github.com/nexus-rpc/sdk-go/nexus"
+	commonpb "go.temporal.io/api/common/v1"
+)
+
+type payloadSerializer struct{}
+
+var errSerializer = errors.New("serializer error")
+
+// Deserialize implements nexus.Serializer.
+func (payloadSerializer) Deserialize(content *nexus.Content, v any) error {
```

Comment:
To me this reads clearer as a future maintainer, but don't have to accept it (just typed real quick, so probably some typos):

```go
// Deserialize implements nexus.Serializer.
func (payloadSerializer) Deserialize(content *nexus.Content, v any) error {
	payloadRef, ok := v.(**commonpb.Payload)
	if !ok {
		return fmt.Errorf("cannot deserialize into %v", v)
	}
	payload := &commonpb.Payload{Metadata: map[string][]byte{}, Data: content.Data}
	*payloadRef = payload

	// Absent content type with no data means null, with data is just binary
	if contentType := content.Header["type"]; contentType == "" {
		if len(content.Data) == 0 {
			payload.Metadata["encoding"] = []byte("binary/null")
		} else {
			payload.Metadata["encoding"] = []byte("binary/plain")
		}
		return nil
	}

	// Different content types make different payloads
	if mediaType, params, err := mime.ParseMediaType(contentType); err == nil {
		switch mediaType {
		case "application/x-temporal-payload":
			if err := payload.Unmarshal(content.Data); err != nil {
				return fmt.Errorf("failed to unmarshal payload: %w", err)
			}
			return nil
		case "application/json":
			if protoMessageType := params["protobuf-message-type"]; protoMessageType != "" {
				payload.Metadata["encoding"] = []byte("json/protobuf")
				payload.Metadata["messageType"] = []byte(protoMessageType)
			} else {
				payload.Metadata["encoding"] = []byte("json/plain")
			}
			return nil
		case "application/x-protobuf":
			if protoMessageType := params["protobuf-message-type"]; protoMessageType != "" {
				payload.Metadata["encoding"] = []byte("binary/protobuf")
				payload.Metadata["messageType"] = []byte(protoMessageType)
				return nil
			}
			return fmt.Errorf("protobuf message type required for application/x-protobuf content type")
		case "application/octet-stream":
			payload.Metadata["encoding"] = []byte("binary/plain")
			return nil
		}
	}

	// Unknown or bad content type
	payload.Metadata["encoding"] = "binary/unknown-nexus"
	for k, v := range content.Header {
		// Don't want length or encoding
		if k != "length" && k != "encoding" {
			payload.Metadata[k] = []byte(v)
		}
	}
	return nil
}
```

Reply:
I prefer my stricter non-lossy version.

Reply:
Thanks for the suggestion though.